packages/go_router/lib/src/redirection.dart

```diff
@@ -0,0 +1,145 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'configuration.dart';
+import 'logging.dart';
+import 'match.dart';
+import 'matching.dart';
+
+/// A GoRouter redirector function.
+// TODO(johnpryan): make redirector async (#105808)
+typedef RouteRedirector = RouteMatchList Function(RouteMatchList matches,
+    RouteConfiguration configuration, RouteMatcher matcher,
+    {Object? extra});
+
+/// Processes redirects by returning a new [RouteMatchList] representing the new
+/// location.
+RouteMatchList redirect(RouteMatchList prevMatchList,
+    RouteConfiguration configuration, RouteMatcher matcher,
+    {Object? extra}) {
+  RouteMatchList matches;
+
+  // Store each redirect to detect loops
+  final List<RouteMatchList> redirects = <RouteMatchList>[prevMatchList];
+
+  // Keep looping until redirecting is done
+  while (true) {
+    final RouteMatchList currentMatches = redirects.last;
+
+    // Check for top-level redirect
+    final Uri uri = currentMatches.location;
+    final String? topRedirectLocation = configuration.topRedirect(
+      GoRouterState(
+        configuration,
+        location: currentMatches.location.toString(),
+        name: null,
+        // No name available at the top level trim the query params off the
+        // sub-location to match route.redirect
+        subloc: uri.path,
+        queryParams: uri.queryParameters,
+        extra: extra,
+      ),
+    );
+
+    // If the new location is null, keep the matches the same as before.
+    if (topRedirectLocation != null) {
+      final RouteMatchList newMatch = matcher.findMatch(topRedirectLocation);
+
+      _addRedirect(redirects, newMatch, prevMatchList.location,
+          configuration.redirectLimit);
+      continue;
+    } else {
+      matches = currentMatches;
+    }
```

Comment:
Nit: what do you think of removing the `else` and moving the comment from line 45 down?

```suggestion
    if (topRedirectLocation != null) {
      final RouteMatchList newMatch = matcher.findMatch(topRedirectLocation);

      _addRedirect(redirects, newMatch, prevMatchList.location,
          configuration.redirectLimit);
      continue;
    }

    // If there's no top-level redirect, keep the matches the same as before.
    matches = currentMatches;
```

Reply:
SGTM